src/box/box.cc

```diff
@@ -641,18 +646,39 @@ wal_stream_abort(struct wal_stream *stream)
  * checking even in release.
  */
 static bool
-wal_stream_has_tx(const struct wal_stream *stream)
+wal_stream_has_tx_in_progress(const struct wal_stream *stream)
 {
 	bool has = stream->tsn != 0;
 	assert(has == (in_txn() != NULL));
 	return has;
 }
 
+/**
+ * The function checks that the tail of the transaction from the journal
+ * has been read.
+ */
+static bool
+wal_stream_has_unfinished_tx(const struct wal_stream *stream)
+{
+	bool has = wal_stream_has_tx_in_progress(stream);
+
+	bool unfinished = false;
+	const struct rlist *nodes_rows = stream->nodes_rows;
+	for (int i = 0; i < VCLOCK_MAX; i++) {
+		if (!rlist_empty((struct rlist *)&nodes_rows[i])) {
+			unfinished = true;
+			break;
+		}
+	}
+
+	return has || unfinished;
```

Comment:
When you see `has` as `true`, you don't need to walk the rlists:
```Diff
@@ -660,18 +660,14 @@ wal_stream_has_tx_in_progress(const struct wal_stream *stream)
 static bool
 wal_stream_has_unfinished_tx(const struct wal_stream *stream)
 {
-	bool has = wal_stream_has_tx_in_progress(stream);
-
-	bool unfinished = false;
+	if (wal_stream_has_tx_in_progress(stream))
+		return true;
 	const struct rlist *nodes_rows = stream->nodes_rows;
 	for (int i = 0; i < VCLOCK_MAX; i++) {
-		if (!rlist_empty((struct rlist *)&nodes_rows[i])) {
-			unfinished = true;
-			break;
-		}
+		if (!rlist_empty((struct rlist *)&nodes_rows[i]))
+			return true;
 	}
-
-	return has || unfinished;
+	return false;
 }
```

Reply:
This looks better. Fixed.